src/ucp/api/ucp.h

```diff
@@ -1269,8 +1272,17 @@ int ucp_request_is_completed(void *request);
  *
  * @param [in]  request      Non-blocking request to release.
  *
- * @note If the request is not completed yet, it will actually be released when
- *       completed.
+ * This routine marks and potentially releases back to the library the
+ * non-blocking request. If the request is already is already completed or
+ * canceled state it is released the resources associated with the request are
+ * returned back to the library.  If the request in any other stated it is
+ * marked as a "ready to release" and it will be released once it enters
+ * completed and canceled states.
+ *
+ * @todo I think this release semantics is a bit confusing. I would suggest to
```

Comment:
we need to do it to satisfy MPI semantics. you can say that we only release the handle here, and do not affect the course of the operation itself.